crypto/storage/client.go

```diff
@@ -0,0 +1,146 @@
+// This server can either be a secret store itself, or proxy the request to a key store such as Hashicorp Vault or Azure Key Vault.
+// It allows us to keep the codebase clean and allow other parties to write their own adaptor.
+type APIClient struct {
+	httpClient *httpclient.ClientWithResponses
+}
+
+func (c APIClient) Name() string {
+	return "Crypto"
+}
+
+func (c APIClient) CheckHealth() map[string]core.Health {
+	results := make(map[string]core.Health)
+	response, err := c.httpClient.HealthCheckWithResponse(context.Background())
+	if err != nil {
+		results[StorageAPIConfigKey] = core.Health{Status: core.HealthStatusDown, Details: fmt.Errorf("unable to connect to storage server: %w", err).Error()}
+		return results
+	}
+
+	switch response.StatusCode() {
+	case http.StatusOK:
+		results[StorageAPIConfigKey] = core.Health{Status: core.HealthStatusUp}
+		break
+	case http.StatusServiceUnavailable:
+		results[StorageAPIConfigKey] = core.Health{Status: core.HealthStatusDown, Details: fmt.Sprintf("storage server reports to be unavailable: %d", response.StatusCode())}
+		break
+	default:
+		results[StorageAPIConfigKey] = core.Health{Status: core.HealthStatusUnknown, Details: fmt.Sprintf("unexpected status code from storage server: %d", response.StatusCode())}
+	}
+
+	return results
+}
+
+type APIClientConfig struct {
+	URL string `koanf:"url"`
+}
+
+// NewAPIClient create a new API Client to communicate with a remote storage server.
+func NewAPIClient(u string) (Storage, error) {
+	if _, err := url.ParseRequestURI(u); err != nil {
+		return nil, err
+	}
+	client, _ := httpclient.NewClientWithResponses(u, httpclient.WithHTTPClient(&http.Client{Timeout: httpClientTimeout}))
+	return &APIClient{httpClient: client}, nil
+}
+
+type backendError struct {
+	error httpclient.ErrorResponse
+}
+
+func (r backendError) Error() string {
+	return fmt.Sprintf("remote error: the backend %s returned an error with status %d, title=%s details=%s", r.error.Backend, r.error.Status, r.error.Title, r.error.Detail)
+}
+
+func (c APIClient) GetPrivateKey(kid string) (crypto.Signer, error) {
+	response, err := c.httpClient.LookupSecretWithResponse(context.Background(), kid)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get private key: %w", err)
+	}
+	switch response.StatusCode() {
+	case http.StatusOK:
+		if response.JSON200 == nil {
+			return nil, fmt.Errorf("unable to get private key: no body or wrong content-type")
+		}
+
+		privateKey, err := util.PemToPrivateKey([]byte(response.JSON200.Secret))
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse private key as PEM: %w", err)
+		}
+		return privateKey, nil
+	case http.StatusNotFound:
+		return nil, ErrNotFound
+	case http.StatusBadRequest:
+		if response.JSON400 != nil {
+			return nil, backendError{error: *response.JSON400}
+		}
+		return nil, fmt.Errorf("unable to get private key: server responded with bad-request: %s", string(response.Body))
```

Comment:
probably not the best idea to include the response in the error, since it might be returned to the end-user client (or it might be binary). Maybe just log it?

And I'd say error 400 is also unexpected, so you can just fallback to the default case (and maybe try to parse it as problem response, so you can log it).